lib/puppet/defaults.rb

```diff
@@ -1871,7 +1872,22 @@ def self.initialize_default_settings!(settings)
       :desc     => "Where puppet agent stores the last run report summary in yaml format."
     },
     :lastrunreport =>  {
-      :default  => "$statedir/last_run_report.yaml",
+      :default  => lambda {
+        # binding.pry
+        # route_file = Puppet[:route_file]
+        if Puppet::FileSystem.exist?(Puppet[:route_file])
+          routes = Puppet::Util::Yaml.safe_load_file(Puppet[:route_file], [Symbol])
+          file_type = routes.dig("agent", "report", "terminus")
+          # file_type = routes.dig("agent", "report", "cache")
+          if file_type.nil?
+            "$statedir/last_run_report.#{file_type}" 
+          else
+            "$statedir/last_run_report.yaml"
+          end
+        else
+          "$statedir/last_run_report.yaml"
+        end
```

Comment:
Maybe this can be simplified as below, and I'm not sure about this is the right location to check this
```
if routes.dig("agent", "report", "terminus") != :yaml
  "$statedir/last_run_report.json"
else
  "$statedir/last_run_report.yaml"
end
```

Reply:
I think this is going to be problematic, because if there is a `routes.yaml` for the agent application that configures the cache terminus to JSON, then `puppet agent` will use a different path than `puppet config print lastrunreport` displays. But the former is often used by monitoring/scripts to automate puppet installation/provisioning. Also I think it's possible for the default value of `Puppet[:lastrunsummary]` to be resolved before the application configures the [cache terminus in code](https://github.com/puppetlabs/puppet/blob/de3c3a8d746bb2e2449110977a3adb58b7f7fa2a/lib/puppet/application/agent.rb#L476)